File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    public int getLoad(int[] weights, int capacity)
    {
        int d = 0;
        int curr = 0;
        for(int i : weights)
        {
            curr += i;
            if(curr > capacity)
            {
                curr = i;
                d++;
            }
        }
        return d+1;
    }
    public int shipWithinDays(int[] weights, int days) {
        int low = weights[0];
        int high = weights[0];
        for(int i = 1; i<weights.length; i++)
        {
            low = low > weights[i] ? low : weights[i];
            high += weights[i];
        }
        while(low<=high)
        {
            int mid = (low + high)/2;
            int d = getLoad(weights, mid);
            if(d <= days)
            {
                high = mid - 1;
            }
            else 
            {
                low = mid + 1;
            }
        }
        return low;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int getdays(vector<int>& weights , int mid)
    {
        int k = 0, repo = 0;
        for(int i =0;i<weights.size();i++)
        {
            k += weights[i];
            if(k>mid)
            {
                k = weights[i];
                // i--;
                repo++;
            }
        }
        return repo+1;
    }
    int shipWithinDays(vector<int>& weights, int days) 
    {    
        int low =0 , high =0;
        for(int i =0;i<weights.size();i++)
        {
          low = max(low, weights[i]);
            high += weights[i];
        }
         while(low<=high)
         {
             int mid= ((high + low)/2);
             int out = getdays(weights , mid);

             if(out>days)
             {
                 low= mid+1;
             }
             else if(out<=days)
             {
                 high= mid-1;

             }
             // else
             // {
             //     high = mid;
             // }


         }
     return low;
         
    }
};